<dom-module id="genome-browser">

    <template>
        <div id="{{prefix}}GenomeBrowser"></div>
    </template>

    <script>
        class GenomeBrowserComponent extends Polymer.Element {

            constructor() {
                super();

                if (UtilsNew.isEmpty(this.prefix)) {
                    this.prefix = "gb_" + Utils.randomString(6);
                }
            }

            static get is() {
                return 'genome-browser';
            }

            static get properties() {
                return {
                    opencgaSession: {
                        type: Object
                    },
                    cellbaseClient: {
                        type: Object
//                        observer: "init"
                    },
                    opencgaClient: {
                        type: Object
                    },
                    region: {
                        type: Object,
                        observer: "_setRegion"
                    },
                    species: {
                        type: Object,
                        value: {
                            id: "hsapiens",
                            scientificName: "Homo sapiens",
                            assembly: {name: "GRCh37"}
                        }
                    },
                    tracks: {
                        type: Object,
                        value: {
                            sequence: {type: "sequence"},
                            gene: {type: "gene"},
                            variant: {type: "variant", config: {}},
                            alignment: {type: "alignment", config: {}}
                        },
                        observer: "init"
                    },
                    width: {
                        type: Number,
//                        value: 1400,
                        observer: "_updateWidth"
                    },
                    active: {
                        type: Boolean,
                        value: false,
                        observer: "_setActive"
                    },
                    settings: {
                        type: Object
                    }
                };
            }

            ready() {
                super.ready();

                this.init();
//                this.region = {chromosome: "21", start: 22859462, end: 22859610};

                // We need to listen to Window resize
                window.addEventListener('resize', e => this._updateWidth());
            }

            _attachDom(dom) {
                this.appendChild(dom);
            }

            init() {
                if (typeof this.cellbaseClient === "undefined" || typeof  this.opencgaSession === "undefined"
                    || typeof this.opencgaClient === "undefined" || typeof this.region === "undefined") {
                    return;
                }

                if (!this.active) {
                    return;
                }

                if (typeof this.genomeBrowser !== "undefined") {
                    this.genomeBrowser.destroy();
                }

                this.width = $("#" + this.prefix + "GenomeBrowser").width();

                let defaultSettings = {
                    client: this.cellbaseClient,
                    target: this.prefix + "GenomeBrowser",
                    region: this.region,
                    availableSpecies: {vertebrates: [this.species]},
                    species: this.species,
                    width: this.width,
//                    zoom: 80,
                    sidePanel: false,
                    autoRender: true,
                    resizable: true,
                    karyotypePanelConfig: {
                        collapsed: true,
                        collapsible: true
                    },
                    chromosomePanelConfig: {
                        collapsed: true,
                        collapsible: true
                    },
                    navigationBarConfig: {
                        componentsConfig: {}
                    },
                    handlers: {
                        'region:change': function (e) {
                            console.log(e)
                        }
                    },
                };

                let settings = Object.assign(defaultSettings, this.settings);
                this.genomeBrowser = new GenomeBrowser(settings);

                let _tracks = [];
                for (let track of Object.values(this.tracks)){
                    switch (track.type) {
                        case 'sequence':
                            _tracks.push(this._addSequenceTrack(track.config));
                            break;
                        case 'gene':
                            _tracks.push(this._addGeneTrack(track.config));
                            break;
//                            case 'snp':
                        case 'variant':
                            _tracks.push(this._addVariantTrack(track.config));
                            break;
                        case 'alignment':
                            // One track per file
                            let alignmentTracks = this.addAlignmentTrack(track.config);
                            if (UtilsNew.isNotEmptyArray(alignmentTracks)) {
                                for (let alignmentTrack of alignmentTracks) {
                                    _tracks.push(alignmentTrack);
                                }
                            }
//                                if (typeof this.samples !== "undefined" && this.samples.length > 0) {
//                                    tracks.push(this._addAlignmentTrack(this.trackConfigs[this.tracks[i]]));
//                                }
                            break;
                        case 'template':
                            _tracks.push(this._addTemplateTrack(track.config));
                            break;
//                            default:
//                                tracks.push(this.addGenericTrack(this.trackConfigs[this.tracks[i]]));
//                                break;
                    }
                }

                // Region Overview
                let gene = new FeatureTrack({
                    minHistogramRegionSize: 20000000,
                    maxLabelRegionSize: 10000000,
                    height: 120,
                    renderer: new FeatureRenderer(FEATURE_TYPES.gene),
                    dataAdapter: new CellBaseAdapter(this.cellbaseClient, "genomic", "region", "gene",
                            {
                                exclude: 'transcripts.tfbs,transcripts.xrefs,transcripts.exons.sequence'
                            }, {
                                chunkSize: 100000
                            })
                });
                this.genomeBrowser.addOverviewTrack(gene);
                this.genomeBrowser.addTrack(_tracks);
                this.genomeBrowser.draw();
            }

            _addSequenceTrack(settings) {
                let mySettings = {
                    title: 'Sequence',
                    height: 20,
                    visibleRegionSize: 200,

                    renderer: new SequenceRenderer(),
                    dataAdapter: new CellBaseAdapter(this.cellbaseClient, "genomic", "region", "sequence", {}, {chunkSize: 100})
                };
                mySettings = Object.assign(mySettings, settings);
                return new FeatureTrack(mySettings);
            }

            _addGeneTrack(settings) {
                let mySettings = {
                    title: 'Gene',
                    minHistogramRegionSize: 20000000,
                    maxLabelRegionSize: 10000000,
                    minTranscriptRegionSize: 200000,
                    height: 100,

                    renderer: new GeneRenderer({
                        handlers: {
                            'feature:click': function (e) {
                                console.log(e)
                            }
                        }
                    }),
                    dataAdapter: new CellBaseAdapter(this.cellbaseClient, "genomic", "region", "gene",
                            {
                                exclude: 'transcripts.tfbs,transcripts.xrefs,transcripts.exons.sequence'
                            }, {
                                chunkSize: 100000
                            })
                };
                mySettings = Object.assign(mySettings, settings);
                return new GeneTrack(mySettings);
            }

            _addVariantTrack(settings) {
//                console.log("Study credentials: " + this.project.alias + ":" + this.study.alias);
                let mySettings = {
                    title: "Variant",
                    closable: true,
                    minHistogramRegionSize: 20000000,
                    maxLabelRegionSize: 10000000,
                    minTranscriptRegionSize: 200000,
                    visibleRegionSize: 100000000,
                    height: 100,
                    opencga: {
                        client: this.opencgaClient,
                        studies: `${this.opencgaSession.project.alias}:${this.opencgaSession.study.alias}`,
                        config: {}
//                        studies: "reference_grch37:1kG_phase3"
//                        "HG00096", "HG00097", "HG00099"
                    },
//                    renderer: new FeatureRenderer(FEATURE_TYPES.variant),
                };

                if (settings.samples !== undefined && settings.samples.length > 0) {
                    let _samples = [];
                    for (let s of settings.samples) {
                        _samples.push(s.name);
                    }

                    mySettings.opencga.samples = _samples;
                } else {
                    mySettings.opencga.samples = [];
                }

                if (settings.files !== undefined && settings.files.length > 0) {
                    let _files = [];
                    for (let s of settings.files) {
                        _files.push(s.name);
                    }

                    mySettings.opencga.files = _files;
                } else {
                    mySettings.opencga.files = [];
                }

                //mySettings = Object.assign(mySettings, settings);
                return new VariantTrack(mySettings);
            }


            getAlignmentTrack(settings) {
                let title = "Aligment";
                if (UtilsNew.isNotUndefinedOrNull(settings.data.name)) {
                    title = `Aligment ${settings.data.name}`;
                }
                let mySettings = {
                    title: title,
                    closable: true,
                    minHistogramRegionSize: 5000,
                    maxLabelRegionSize: 3000,
                    visibleRegionSize: 100000000,
                    height: 300,
                    opencga: {
                        client: this.opencgaClient,
                        studies: `${this.opencgaSession.project.alias}:${this.opencgaSession.study.alias}`,
//                        config: {height:1},
                        // By default no custom config
                        config: {},

//                        studies: "reference_grch37:1kG_phase3"
//                        "HG00096", "HG00097", "HG00099"
                    }
                };

                // You have to pass opencga.config to renderer if you use your own renderer. If you customize config yo have to change opencga.config.
                let renderer = {
                    renderer: new AlignmentRenderer({config: mySettings.opencga.config}),
                    dataAdapter: new OpencgaAdapter(this.opencgaClient, "analysis/alignment", undefined, "query", settings.data)
                };
                mySettings = Object.assign(mySettings, settings.config, renderer);
                return new AlignmentTrack(mySettings);
            }

            /*
             settings is supposed to have:
             - type: alignment...
             - config: Configuration of the track (title, closable, height...)
             - source: opencga, cellbase...
             - data: data that will be used
             */
            addAlignmentTrack(settings = {}) {
                let resTracks = [];

                if (UtilsNew.isNotUndefinedOrNull(settings.files) && settings.files.length > 0) {
                    let _this = this;
                    settings.files.forEach(function (file) {
                        let mySettings = {
                            data: {
                                fileId: file.path,
                                name: file.name,
                                study: _this.opencgaSession.study.id
                            }
                        };

                        // if (UtilsNew.isNotEmptyArray(settings.filesName) && settings.filesName.length > 0) {
                        //     mySettings.data.name = settings.filesName[file];
                        //
                        // }

                        Object.assign(settings, mySettings);
                        let track = _this.getAlignmentTrack(settings);
                        resTracks.push(track);

                    });
                }
                return resTracks
            }

            addGenericTrack(settings) {
                switch (settings.type) {
                    case 'alignment':
                    default:
                        return this.getAlignmentTrack(settings);
                }
            }

            /* Building
              example:
                uriTemplate: "http://ws.babelomics.org/cellbase/webservices/rest/v4/hsapiens/genomic/region/13:32990000-32999999/snp?exclude={exclude}"
                templateVariables: {exclude:'annotation.populationFrequencies,annotation.additionalAttributes,transcriptVariations,xrefs,samples'}
             */
            _addTemplateTrack(settings = {}) {

                let uriTemplate = settings.uriTemplate;
                let templateVariables = settings.templateVariables;
                let title = settings.title;
                let t = new FeatureTrack({
                    title: title,
                    closable: true,
                    minHistogramRegionSize: 20000000,
                    maxLabelRegionSize: 10000000,
                    minTranscriptRegionSize: 200000,
                    histogramColor: '#92de47',
                    height: 100,
                    renderer: new FeatureRenderer(FEATURE_TYPES.snp),
                    dataAdapter: new FeatureTemplateAdapter({
                        multiRegions: false,
                        histogramMultiRegions: false,
                        uriTemplate: uriTemplate,
                        templateVariables: templateVariables,
                        species: genomeBrowser.species,
                        cacheConfig: {
                            chunkSize: 100000
                        },
                        parse: function(response) {
                            var chunks = [];
                            for(var i = 0; i< response.response.length; i++){
                                var res = response.response[i].result;
                                chunks.push(res);
                            }
                            return chunks;
                        }
                    })
                });
                return(t);
            }

            _setRegion(region) {
                if (typeof this.genomeBrowser === "undefined") {
                    this.init();
                }

                if (typeof this.genomeBrowser !== "undefined" && this.active) {
                    this.genomeBrowser.setRegion(region);
                }
            }

//            _setSamples(samples) {
//                if (typeof this.genomeBrowser === "undefined") {
//                    this.init();
//                }
//            }

            _updateWidth(width) {
                if (typeof this.genomeBrowser !== "undefined") {
                    // Check if width has been passed, otherwise we calculate the max for the current DIV size
                    if (width === undefined || width === null || width < 0) {
                        width = $("#" + this.prefix + "GenomeBrowser").width();
                    }

                    // Set the with to the genome browser
                    this.genomeBrowser.setWidth(width);
                }
            }

            _setActive(active) {
                if (active && typeof this.genomeBrowser === "undefined") {
                    this.init();
                }
            }

        }

        customElements.define(GenomeBrowserComponent.is, GenomeBrowserComponent);
    </script>
</dom-module>
